### ISSUE TITLE
Loss Implementation - II

### ISSUE THREAD

**Issue by toshi2k2** [stefanopini/simple-HRNet#29]
Just a followup from the last issue("Loss Implementation") which was closed. I just checked the saved weights of the HR Net models but it turns out that they have the batchnorm parameters as well (running_mean, etc). Also, the adam parameters are available in the yaml files on the actual repo (the lr's 1e-5, etc). But, continuing retraining from the saved model still seems to drop the accuracy to 0. Any idea why that may be happening?

**Comment by stefanopini** [comment]
Can you please provide more details?
I tried continuing the training for a couple of epochs starting from official weights `pose_hrnet_w32_256x192.pth` and using `lr=1e-5` and it worked fine.
If you want to use the adam parameters of the yamls files, you can change them here: https://github.com/stefanopini/simple-HRNet/blob/master/training/Train.py#L177

Btw, you are right, batch norm parameters are included into the weights. However, the state of the Adam optimizer is not (its _trainable_ parameters).
The Adam optimizer creates "exponential moving averages of gradient values" and updates them during the training. Restarting a training without restoring the optimizer will result in empty exponential averages (as at the beginning of the training)... This could (slightly?) affect the results.

**Comment by toshi2k2** [comment]
I'm a bit confused by the Precision values of the retraining model.
![image](https://user-images.githubusercontent.com/38487734/67435988-5e391700-f5bb-11e9-8a5e-de5300c36cb0.png)
Are these values normal?

**Comment by stefanopini** [comment]
I think these results can be plausible because strong data augmentation is applied on training samples while validation accuracy should fluctuate around the initial value or increase.
Actually, I do not recall neither drops nor improvements on the validation results after some training epochs using `lr = 1e-5`, but I'll try again and post the results when I have again the chance to do that.

**Comment by toshi2k2** [comment]
Epoch 2 of 200 @ 2019-10-23 17:29:59
Training: 100%|████████████████████████████████████████████████████████| 9363/9363 [4:15:44<00:00,  1.94s/it]

Train AP/AR
WARNING:root:json-tricks: numpy scalar serialization is experimental and may work differently in future versions
Loading and preparing results...
DONE (t=15.83s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=126.04s).
Accumulating evaluation results...
DONE (t=5.00s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.081
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.137
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.080
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.094
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.072
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.213
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.318
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.221
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.200
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.231
Validating: 100%|██████████████████████████████████████████████████████████| 397/397 [03:53<00:00,  2.53it/s]

Val AP/AR
Loading and preparing results...
DONE (t=0.70s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=6.30s).
Accumulating evaluation results...
DONE (t=0.15s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.692
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.916
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.782
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.662
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.738
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.726
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.928
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.808
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.690
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.780
best_loss 0.000530 at epoch 2
best_acc 0.850155 at epoch 2
best_mAP 0.692310 at epoch 2

Epoch 3 of 200 @ 2019-10-23 22:02:34
Training: 100%|████████████████████████████████████████████████████████| 9363/9363 [4:40:07<00:00,  1.74s/it]

Train AP/AR
Loading and preparing results...
DONE (t=15.99s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=122.63s).
Accumulating evaluation results...
DONE (t=4.77s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.074
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.134
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.071
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.087
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.067
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.193
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.305
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.199
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.180
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.211
Validating: 100%|██████████████████████████████████████████████████████████| 397/397 [03:42<00:00,  2.19it/s]

Val AP/AR
Loading and preparing results...
DONE (t=0.54s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=5.13s).
Accumulating evaluation results...
DONE (t=0.12s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.641
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.894
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.728
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.612
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.685
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.676
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.904
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.754
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.641
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.728

**Comment by stefanopini** [comment]
Thank you very much for posting.

I've just found [here ](https://github.com/leoxiaobin/deep-high-resolution-net.pytorch/blob/master/experiments/coco/hrnet/w32_256x192_adam_lr1e-3.yaml#L98-L108) and [here](https://github.com/leoxiaobin/deep-high-resolution-net.pytorch/blob/master/experiments/coco/hrnet/w48_384x288_adam_lr1e-3.yaml#L98-L108) that authors did not use default Adam settings, but set the following values in the yaml configuration files:

>   WD: 0.0001
>   GAMMA1: 0.99
>   GAMMA2: 0.0
>   MOMENTUM: 0.9
>   NESTEROV: false

However, [here](https://github.com/leoxiaobin/deep-high-resolution-net.pytorch/blob/cb27f9aa4ee338ea70f1660e5146c21356bdffb3/lib/utils/utils.py#L70-L74) it seems that they didn't use these settings... I don't know what to say, maybe they uploaded the wrong code.

Can you try re-running the training script changing [this line](https://github.com/stefanopini/simple-HRNet/blob/master/training/Train.py#L177) with:

- `self.optim = Adam(self.model.parameters(), lr=1e-5, weight_decay=0.0001, betas=[0.99, 0.0])  # setting parameters written in yaml configuration files`

and

- `self.optim = Adam(self.model.parameters(), lr=1e-5)  # using the code of the original repository`

and post the validation results for a couple of epochs?

**Comment by toshi2k2** [comment]
For  ```self.optim = Adam(self.model.parameters(), lr=1e-5, weight_decay=0.0001, betas=[0.99, 0.0]) # setting parameters written in yaml configuration files```
![image](https://user-images.githubusercontent.com/38487734/68480659-560be900-0203-11ea-9bbf-2a2075de7622.png)

**Comment by stefanopini** [comment]
Ok, I think that this slightly decrease could be caused by the moving averages of Adam (which are not provided) while bad training accuracy is caused by strong data augmentation.

I'll update the default hyper-parameters of the training code setting `weight_decay=0.`

I think we can close this, please let me know if you have further questions.
